### PR TITLE
OpenMPTarget: Block failing unit tests with clang/16.

### PR DESCRIPTION
**Pull request by rgayatri23**
The PR adds unit tests that need to be blocked for clang/16 with the OpenMPTarget backend.

### CODE REVIEW

**Review by masterleinad — DISMISSED** [review]

**Comment by masterleinad** [comment]
I'm a little surprised. Our CI is currently passing (apart from the `partition_by_*` tests that fail regularly) and we don't see issues with the ones you are disabling here.

**Comment by rgayatri23** [comment]
> I'm a little surprised. Our CI is currently passing (apart from the `partition_by_*` tests that fail regularly) and we don't see issues with the ones you are disabling here.

Maybe the issue is with A100? I am testing this on Perlmutter (on which `partition_by_*` is actually working). I dont have access to any V100 machine. Not sure whats happening. We can also block these based on compiler **&** architecture.

**Comment by dalg24** [comment]
I think Daniel's point is why are we disabling more tests for Clang 16+ when our CI build that has Clang 17 passes w/o these removal

**Comment by rgayatri23** [comment]
> I think Daniel's point is why are we disabling more tests for Clang 16+ when our CI build that has Clang 17 passes w/o these removal

Ok that makes sense. I can close this PR if you think we don't need to add this additional blockings.

**Comment by rgayatri23** [comment]
Edited the CMakeLists to block tests based on strict compiler versions. It looks like from clang/17.0.2 onwards all regressions have been fixed.

**Comment by masterleinad** [comment]
> Ok that makes sense. I can close this PR if you think we don't need to add this additional blockings.

I'm fine with these extra guards if it helps you with working on the `OpenMPTarget` backend. Otherwise, I would not bother with it. The unit tests are mostly used in our CI.

File: algorithms/unit_tests/CMakeLists.txt [hunk 1]
```diff
@@ -351,7 +351,7 @@ endforeach()
 # FIXME_OPENMPTARGET - remove sort test as it leads to ICE with clang/16 and above at compile time.
 if(KOKKOS_ENABLE_OPENMPTARGET AND KOKKOS_CXX_COMPILER_ID STREQUAL "Clang" AND KOKKOS_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 16.0.0)
     list(REMOVE_ITEM ALGO_SORT_SOURCES
-    TestSort.cpp
+    ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestSort.cpp
```
Comment (Rombur) [inline algorithms/unit_tests/CMakeLists.txt:354]: Why did you change this? The code was wrong before?

File: core/unit_test/CMakeLists.txt [hunk 2]
```diff
@@ -372,20 +372,22 @@ if(Kokkos_ENABLE_OPENMPTARGET)
     ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_TeamCombinedReducers.cpp
     ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_TeamReductionScan.cpp
     ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_WorkGraph.cpp
-    IF (KOKKOS_CXX_COMPILER_ID STREQUAL "Clang" AND KOKKOS_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 16.0.0)
+    IF (KOKKOS_CXX_COMPILER_ID STREQUAL "Clang" AND KOKKOS_CXX_COMPILER_VERSION VERSION_EQUAL 16.0.0)
```
Comment (Rombur) [inline core/unit_test/CMakeLists.txt:375]: Shouldn't we check for 16.X instead of only 16.0.0?

**Comment by rgayatri23** [comment]
> > Ok that makes sense. I can close this PR if you think we don't need to add this additional blockings.
> 
> I'm fine with these extra guards if it helps you with working on the `OpenMPTarget` backend. Otherwise, I would not bother with it. The unit tests are mostly used in our CI.

Well this PR enables more unit tests with 17.0.1 (which is in our CI) compared to what we have now.

Alternatively if we can think of updating our CI to 17.0.2, we can enable all unit tests that were working with clang/15.

File: algorithms/unit_tests/CMakeLists.txt (hunk 1 above)
Comment (rgayatri23) [reply algorithms/unit_tests/CMakeLists.txt:354]: Yes, it was wrong before. We did not catch it since the CI was also updated to 17.0.1

File: core/unit_test/CMakeLists.txt (hunk 2 above)
Comment (rgayatri23) [reply core/unit_test/CMakeLists.txt:375]: Good point. I can do that if we are fine with additional guards that don't really add anything to the CI .

**Comment by masterleinad** [comment]
For reference, with #6511 and `clang 17.0.2` I get the tests to pass with
```diff
diff --git a/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp b/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp
index fb75f05f2..6589b989a 100644
--- a/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp
+++ b/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp
@@ -619,8 +619,8 @@ struct ParallelReduceSpecialize<FunctorType, TeamPolicyInternal<PolicyArgs...>,
 // exit.
 #pragma omp target map(to : final_reducer) is_device_ptr(scratch_ptr)
       {
-        final_reducer.init(scratch_ptr);
-        final_reducer.final(scratch_ptr);
+        final_reducer.init(static_cast<PointerType>(scratch_ptr));
+        final_reducer.final(static_cast<PointerType>(scratch_ptr));
       }
 
       if (0 < value_count) {
@@ -680,7 +680,7 @@ struct ParallelReduceSpecialize<FunctorType, TeamPolicyInternal<PolicyArgs...>,
         // Do the final only once at the end.
         if (tree_neighbor_offset * 2 >= nteams && omp_get_team_num() == 0 &&
             omp_get_thread_num() == 0) {
-          final_reducer.final(scratch_ptr);
+          final_reducer.final(static_cast<PointerType>(scratch_ptr));
         }
       }
       tree_neighbor_offset *= 2;
diff --git a/core/unit_test/CMakeLists.txt b/core/unit_test/CMakeLists.txt
index b71c72c3c..c06796e92 100644
--- a/core/unit_test/CMakeLists.txt
+++ b/core/unit_test/CMakeLists.txt
@@ -368,17 +368,17 @@ endforeach()
 # Disable non-compiling tests based on clang version.
 if(Kokkos_ENABLE_OPENMPTARGET)
   list(REMOVE_ITEM OpenMPTarget_SOURCES
-    ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_Other.cpp
+	  #${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_Other.cpp
     ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_TeamCombinedReducers.cpp
     ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_TeamReductionScan.cpp
     ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_WorkGraph.cpp
-    IF (KOKKOS_CXX_COMPILER_ID STREQUAL "Clang" AND KOKKOS_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 16.0.0)
+    #IF (KOKKOS_CXX_COMPILER_ID STREQUAL "Clang" AND KOKKOS_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 16.0.0)
         ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_SubView_c01.cpp
         ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_SubView_c02.cpp
         ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_SubView_c03.cpp
         ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_Reducers_d.cpp
-    endif()
-    IF (KOKKOS_CXX_COMPILER_ID STREQUAL "Clang" AND KOKKOS_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 16.0.0)
+	#endif()
+	#IF (KOKKOS_CXX_COMPILER_ID STREQUAL "Clang" AND KOKKOS_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 16.0.0)
         ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_AtomicOperations_shared.cpp
         ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_MinMaxClamp.cpp
         ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_TeamVectorRange.cpp
@@ -386,7 +386,7 @@ if(Kokkos_ENABLE_OPENMPTARGET)
         ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_TeamScan.cpp
         ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_TeamBasic.cpp
         ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_ViewAPI_e.cpp
-    endif()
+	#endif()
     # FIXME_OPENMPTARGET_CRAY: The following tests fail at compile time when the OpenMPTarget backend is enabled with the Cray compiler.
     # Atomic compare/exchange is used in these tests which can be one of the reasons for the compilation failures.
     IF(KOKKOS_CXX_COMPILER_ID STREQUAL Cray)
diff --git a/core/unit_test/TestCXX11.hpp b/core/unit_test/TestCXX11.hpp
index b35de9a3a..bd093e5fb 100644
--- a/core/unit_test/TestCXX11.hpp
+++ b/core/unit_test/TestCXX11.hpp
@@ -347,6 +347,10 @@ TEST(TEST_CATEGORY, cxx11) {
     ASSERT_TRUE((TestCXX11::Test<TEST_EXECSPACE>(1)));
     ASSERT_TRUE((TestCXX11::Test<TEST_EXECSPACE>(2)));
     ASSERT_TRUE((TestCXX11::Test<TEST_EXECSPACE>(3)));
+#ifdef KOKKOS_ENABLE_OPENMPTARGET
+    if (std::is_same_v<TEST_EXECSPACE, Kokkos::Experimental::OpenMPTarget>)
+      GTEST_SKIP() << "Wrong results with OpenMPTarget";
+#endif
     ASSERT_TRUE((TestCXX11::Test<TEST_EXECSPACE>(4)));
   }
 }
diff --git a/core/unit_test/TestConcepts.hpp b/core/unit_test/TestConcepts.hpp
index 476a88483..19add9a0e 100644
--- a/core/unit_test/TestConcepts.hpp
+++ b/core/unit_test/TestConcepts.hpp
@@ -175,9 +175,6 @@ inline constexpr bool is_team_handle_complete_trait_check_v =
 // actual test begins here
 
 /*
-  FIXME_OPENMPTARGET
-  https://github.com/kokkos/kokkos/blob/2d6cbad7e079eb45ae69ac6a59929d9fcf10409a/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp#L860-L864
-
   FIXME_OPENACC
   OpenACCTeamMember is missing the following method:
     template <typename ReducerType>
@@ -185,7 +182,7 @@ inline constexpr bool is_team_handle_complete_trait_check_v =
     team_reduce(ReducerType const& reducer) const noexcept;
 */
 
-#if !defined(KOKKOS_ENABLE_OPENMPTARGET) && !defined(KOKKOS_ENABLE_OPENACC)
+#if !defined(KOKKOS_ENABLE_OPENACC)
 using space_t  = TEST_EXECSPACE;
 using policy_t = Kokkos::TeamPolicy<space_t>;
 using member_t = typename policy_t::member_type;
diff --git a/core/unit_test/TestTeamBasic.hpp b/core/unit_test/TestTeamBasic.hpp
index c395bc083..e41b3fad9 100644
--- a/core/unit_test/TestTeamBasic.hpp
+++ b/core/unit_test/TestTeamBasic.hpp
@@ -37,8 +37,6 @@ TEST(TEST_CATEGORY, team_for) {
       1000);
 }
 
-// FIXME_OPENMPTARGET wrong results
-#ifndef KOKKOS_ENABLE_OPENMPTARGET
 TEST(TEST_CATEGORY, team_reduce) {
   TestTeamPolicy<TEST_EXECSPACE,
                  Kokkos::Schedule<Kokkos::Static> >::test_reduce(0);
@@ -53,7 +51,6 @@ TEST(TEST_CATEGORY, team_reduce) {
   TestTeamPolicy<TEST_EXECSPACE,
                  Kokkos::Schedule<Kokkos::Dynamic> >::test_reduce(1000);
 }
-#endif
 
 template <typename ExecutionSpace>
 struct TestTeamReduceLarge {
@@ -402,7 +399,7 @@ TEST(TEST_CATEGORY, team_handle_by_value) {
 
 }  // namespace Test
 
-#ifndef KOKKOS_ENABLE_OPENMPTARGET
+//#ifndef KOKKOS_ENABLE_OPENMPTARGET
 #include <TestTeamVector.hpp>
-#endif
+//#endif
 #endif
diff --git a/core/unit_test/TestTeamReductionScan.hpp b/core/unit_test/TestTeamReductionScan.hpp
index 8bdd5e843..d6016c545 100644
--- a/core/unit_test/TestTeamReductionScan.hpp
+++ b/core/unit_test/TestTeamReductionScan.hpp
@@ -30,10 +30,6 @@ TEST(TEST_CATEGORY, team_reduction_scan) {
 }
 
 TEST(TEST_CATEGORY, team_long_reduce) {
-#ifdef KOKKOS_ENABLE_OPENMPTARGET  // FIXME_OPENMPTARGET: Not implemented
-  if constexpr (!std::is_same<TEST_EXECSPACE,
-                              Kokkos::Experimental::OpenMPTarget>::value)
-#endif
   {
     TestReduceTeam<long, TEST_EXECSPACE, Kokkos::Schedule<Kokkos::Static> >(0);
     TestReduceTeam<long, TEST_EXECSPACE, Kokkos::Schedule<Kokkos::Dynamic> >(0);
@@ -47,10 +43,6 @@ TEST(TEST_CATEGORY, team_long_reduce) {
 }
 
 TEST(TEST_CATEGORY, team_double_reduce) {
-#ifdef KOKKOS_ENABLE_OPENMPTARGET  // FIXME_OPENMPTARGET: Not implemented
-  if constexpr (!std::is_same<TEST_EXECSPACE,
-                              Kokkos::Experimental::OpenMPTarget>::value)
-#endif
   {
     TestReduceTeam<double, TEST_EXECSPACE, Kokkos::Schedule<Kokkos::Static> >(
         0);
@@ -96,10 +88,6 @@ void test_team_parallel_reduce(const int num_loop_size) {
 }
 
 TEST(TEST_CATEGORY, team_parallel_dummy_with_reducer_and_scratch_space) {
-#ifdef KOKKOS_ENABLE_OPENMPTARGET  // FIXME_OPENMPTARGET: Not implemented
-  if constexpr (!std::is_same<TEST_EXECSPACE,
-                              Kokkos::Experimental::OpenMPTarget>::value)
-#endif
   {
     test_team_parallel_reduce<TEST_EXECSPACE>(0);
     test_team_parallel_reduce<TEST_EXECSPACE>(1);
@@ -107,12 +95,6 @@ TEST(TEST_CATEGORY, team_parallel_dummy_with_reducer_and_scratch_space) {
 }
 
 TEST(TEST_CATEGORY, repeated_team_reduce) {
-#ifdef KOKKOS_ENABLE_OPENMPTARGET
-  if (std::is_same<TEST_EXECSPACE, Kokkos::Experimental::OpenMPTarget>::value)
-    GTEST_SKIP() << "skipping since team_reduce for OpenMPTarget is not "
-                    "properly implemented";
-#endif
-
 #ifdef KOKKOS_IMPL_32BIT
   GTEST_SKIP() << "Failing KOKKOS_IMPL_32BIT";  // FIXME_32BIT
 #endif
```

**Comment by rgayatri23** [comment]
Right, clang/17.0.2 fixes a lot of issues. Should we just update the CI to that?

**Comment by masterleinad** [comment]
> Right, clang/17.0.2 fixes a lot of issues. Should we just update the CI to that?

I'm fine with that.

**Comment by rgayatri23** [comment]
> > Right, clang/17.0.2 fixes a lot of issues. Should we just update the CI to that?
> 
> I'm fine with that.

Should we still protect against 16,17.0.1 ?

**Comment by masterleinad** [comment]
> Should we still protect against 16,17.0.1 ?

I don't see much of a reason to have extra guards for tests but if we know that certain features only work with sufficiently new compiler versions we should have guards for that (in the actual implementation).
It would also be good to decide on the minimum compiler versions we support for `OpenMPTarget` so that we can drop guards for lower versions.

**Comment by rgayatri23** [comment]
> > Should we still protect against 16,17.0.1 ?
> 
> I don't see much of a reason to have extra guards for tests but if we know that certain features only work with sufficiently new compiler versions we should have guards for that (in the actual implementation). It would also be good to decide on the minimum compiler versions we support for `OpenMPTarget` so that we can drop guards for lower versions.

Right, the idea was to have the minimum compiler version requirement in a different PR since then we can pinpoint to that in case that creates issues for someone. I guess we can close this PR and start a separate PR which updates the CI and specifies minimum version and that can also remove guards around unit tests for older compiler versions.

**Comment by masterleinad** [comment]
@rgayatri23 What's the status here?

**Comment by rgayatri23** [comment]
I think we can close this in favor of #6472
